### ISSUE TITLE
Race condition in heartbeat

### ISSUE THREAD

**Issue by Artimi** [aio-libs/aiokafka#88]
During testing our code we've tweaked some parameters of `AIOKafkaConsumer`. One of them is `heartbeat_interval_ms` that we set to 100 ms. After this we run a test that was waiting 100 ms between subscribing to more topics. Strange thing occurred sometimes (at first 1 of 20 cases, then 1 of 200 cases) that when I subscribed and unsubscribed a topic too fast `kafka/aiokafka` raised this exception:
```
    raise self._exception
/usr/lib/python3.5/asyncio/tasks.py:239: in _step
    result = coro.send(None)
messaging_client/mqclient/kafka_consumer.py:227: in stop
    await self._consumer.stop()
../../.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/consumer.py:279: in stop
    yield from self._coordinator.close()
../../.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py:132: in close
    yield from self.heartbeat_task
/usr/lib/python3.5/asyncio/futures.py:363: in __iter__
    return self.result()  # May raise too.
/usr/lib/python3.5/asyncio/futures.py:274: in result
    raise self._exception
/usr/lib/python3.5/asyncio/tasks.py:239: in _step
    result = coro.send(None)
../../.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py:789: in _heartbeat_task_routine
    yield from self.ensure_active_group()
../../.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py:593: in ensure_active_group
    yield from self._perform_group_join()
../../.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py:677: in _perform_group_join
    self.protocol, member_assignment_bytes)
../../.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py:266: in _on_join_complete
    self._subscription.assign_from_subscribed(assignment.partitions())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <kafka.consumer.subscription_state.SubscriptionState object at 0x7f62fcb32240>, assignments = [TopicPartition(topic='TestSchema.c6d8cc1a2fb242b38bee95a4a189e1fa.1', partition=0)]

    def assign_from_subscribed(self, assignments):
        """Update the assignment to the specified partitions
    
            This method is called by the coordinator to dynamically assign
            partitions based on the consumer's topic subscription. This is different
            from assign_from_user() which directly sets the assignment from a
            user-supplied TopicPartition list.
    
            Arguments:
                assignments (list of TopicPartition): partitions to assign to this
                    consumer instance.
            """
        if self.subscription is None:
            raise IllegalStateError(self._SUBSCRIPTION_EXCEPTION_MESSAGE)
    
        for tp in assignments:
            if tp.topic not in self.subscription:
>               raise ValueError("Assigned partition %s for non-subscribed topic." % str(tp))
E               ValueError: Assigned partition TopicPartition(topic='TestSchema.c6d8cc1a2fb242b38bee95a4a189e1fa.1', partition=0) for non-subscribed topic.
``` 
I tried to capture some logs and will show important part of it:
```
[1480672858.4598098][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Received successful heartbeat response.
[1480672858.5570168][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.consumer.subscription_state][INFO] Updating subscribed topics to: {'TestSchema.aaa34fbd914a4d90ad20335be9061c77.1', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.3'}
[1480672858.557299][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.consumer][DEBUG] Subscribed to topic(s): {'TestSchema.aaa34fbd914a4d90ad20335be9061c77.1', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.3'}
[1480672858.558063][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka][DEBUG] Sending metadata request MetadataRequest(topics=['TestSchema.aaa34fbd914a4d90ad20335be9061c77.1', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.3']) to 0
[1480672858.559933][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Revoking previously assigned partitions set()
[1480672858.560181][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] (Re-)joining group 794401da87a540daba226dee38a81b1c
[1480672858.5605147][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Issuing request (JoinGroupRequest(group='794401da87a540daba226dee38a81b1c', session_timeout=30000, member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', protocol_type='consumer', group_protocols=[(protocol_name='roundrobin', protocol_metadata=b'\x00\x00\x00\x00\x00\x03\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.1\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.2\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.3\x00\x00\x00\x00')])) to coordinator 0
[1480672858.669306][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.consumer.subscription_state][INFO] Updating subscribed topics to: {'TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.3'}
[1480672858.6696045][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.consumer][DEBUG] Subscribed to topic(s): {'TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.3'}
[1480672858.760019][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.conn][DEBUG] <AIOKafkaConnection host=10.107.20.205 port=9092> Response 7: MetadataResponse(brokers=[(node_id=0, host='10.107.20.205', port=9092)], topics=[(error_code=0, topic='TestSchema.aaa34fbd914a4d90ad20335be9061c77.1', partitions=[(error_code=0, partition=0, leader=0, replicas=[0], isr=[0])]), (error_code=5, topic='TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', partitions=[]), (error_code=5, topic='TestSchema.aaa34fbd914a4d90ad20335be9061c77.3', partitions=[])])
[1480672858.7602575][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.cluster][WARNING] Topic TestSchema.aaa34fbd914a4d90ad20335be9061c77.2 is not available during auto-create initialization
[1480672858.7603266][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.cluster][WARNING] Topic TestSchema.aaa34fbd914a4d90ad20335be9061c77.3 is not available during auto-create initialization
[1480672858.773429][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.cluster][DEBUG] Updated cluster metadata to Cluster(brokers: 1, topics: 1, groups: 0)
[1480672858.7737765][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.conn][DEBUG] <AIOKafkaConnection host=10.107.20.205 port=9092> Response 8: JoinGroupResponse(error_code=0, generation_id=2, group_protocol='roundrobin', leader_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', members=[(member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', member_metadata=b'\x00\x00\x00\x00\x00\x03\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.1\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.2\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.3\x00\x00\x00\x00')])
[1480672858.7740512][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Join group response JoinGroupResponse(error_code=0, generation_id=2, group_protocol='roundrobin', leader_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', members=[(member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', member_metadata=b'\x00\x00\x00\x00\x00\x03\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.1\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.2\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.3\x00\x00\x00\x00')])
[1480672858.7741585][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][INFO] Joined group '794401da87a540daba226dee38a81b1c' (generation 2) with member_id aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa
[1480672858.774216][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][INFO] Elected group leader -- performing partition assignments using roundrobin
[1480672858.7743132][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Performing roundrobin assignment for subscriptions {'aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa': ConsumerProtocolMemberMetadata(version=0, subscription=['TestSchema.aaa34fbd914a4d90ad20335be9061c77.1', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.2', 'TestSchema.aaa34fbd914a4d90ad20335be9061c77.3'], user_data=b'')}
[1480672858.7743835][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.coordinator.assignors.roundrobin][WARNING] No partition metadata for topic TestSchema.aaa34fbd914a4d90ad20335be9061c77.2
[1480672858.774436][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][kafka.coordinator.assignors.roundrobin][WARNING] No partition metadata for topic TestSchema.aaa34fbd914a4d90ad20335be9061c77.3
[1480672858.774536][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Finished assignment: {'aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa': ConsumerProtocolMemberAssignment(version=0, assignment=[(topic='TestSchema.aaa34fbd914a4d90ad20335be9061c77.1', partitions=[0])], user_data=b'')}
[1480672858.774634][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Issuing leader SyncGroup (SyncGroupRequest(group='794401da87a540daba226dee38a81b1c', generation_id=2, member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', group_assignment=[(member_id='aiokafka-0.1.2-030f6183-dbbb-4733-895a-8f8fe4527dfa', member_metadata=b'\x00\x00\x00\x00\x00\x01\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.1\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00')])) to coordinator 0
[1480672858.7753685][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.conn][DEBUG] <AIOKafkaConnection host=10.107.20.205 port=9092> Response 9: SyncGroupResponse(error_code=0, member_assignment=b'\x00\x00\x00\x00\x00\x01\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.1\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00')
[1480672858.77553][ubuntu][/home/artimi/.virtualenvs/messaging/bin/pytest][aiokafka.group_coordinator][DEBUG] Received successful sync group response for group 794401da87a540daba226dee38a81b1c: SyncGroupResponse(error_code=0, member_assignment=b'\x00\x00\x00\x00\x00\x01\x00-TestSchema.aaa34fbd914a4d90ad20335be9061c77.1\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00')
```

You can see on line 2 that we are subscribing to topics ending with 1, 2 and 3. Then on line 4 `MetadataRequest` is sent. Then on line 8 topics are updated to leave out 1 and only keep 2 and 3. But on line 10 `MetadataResponse` finally arrives with topic 1 as successful and topic 2 and 3 with error 5 (that does not bother us right now). So we are now in situation where we have `subscriptions == {2, 3}`, `assignments == {1}`. Then when kafka method `SubscriptionState.assign_from_subscribed()` is called it discovers that the topic partion 1 is not in list of subscription and raises that previously described error.

Basically, assignment partitions are in state that does not correspond with current subscriptions because metadata are not updated fast (often?) enough. We hopefully found a workaround this by calling `force_metadata_update` after every unsubscription (it is quite uncommon in our workflow so we can afford it).

This happens rarely and unfortunately I could not create reliable minimal working example of this bug. I hope this helps somebody because I spent some time tracking this one.

This happened with `aiokafka == 0.1.2` and kafka in docker container `quantlane/kafka:0.9`.

**Comment by fabregas** [comment]
if I have correctly understood the issue, you change subscription before coordinator join group with "old" assignment.
In this case you can call `ensure_active_group()` method of GroupCoordinator before change subscription. Something like this:

`await consumer._coordinator.ensure_active_group()
consumer.assign(new_partitions)`

**Comment by Drizzt1991** [comment]
Hmm, maybe we should consider to change some API to coroutines for those cases. Basically `assign` will have to:
  * wait until full group is rebalanced
  * we can't do it if a rebalance is in progress, so it should wait for current one to finish.
@Artimi Is it possible to provide a code, that can replicate this? You should also consider upgrading to new `aiokafka`, as we made a bunch of changes from `0.1.2`

**Comment by Artimi** [comment]
@Drizzt1991 Unfortunately I cannot provide any code that will replicate this. We have our own consumers and producers built on top of `aiokafka` and I cannot share it. I tried to run that test with the parameters used before with `aiokafka==0.2.0` and it seems ok but you cannot be sure that it is working reliably as far as it already happened in less than 1% cases. I wanted to report this mainly if somebody encounter this problem. I understand this sort of problem is hard to track and solve. I think it won't cause problems to us with normal heartbeat interval.

**Comment by Drizzt1991** [comment]
@Artimi It's ok, I do see some probable places as it is. Lets let this bug stay here if anyone stumps upon the problem.

**Comment by Artimi** [comment]
Hello again,
this problem popped up again in our environment even though I have used `force_metada_update` after every unsubscription. I tried to replicate it and discovered that the bug shows more often in a fresh new kafka. So when I run a fresh Kafka from docker like this:
```
$ docker run -d -p 9092:9092 -e 'ADVERTISED_HOST=<IP>'  --name kafka quantlane/kafka:0.9
```
Where quantlane/kafka is image defined [here](https://github.com/qntln/docker-kafka/blob/master/kafka/Dockerfile). I can easily achieve this behavior in few tries.

Then I wrote following script that emulates our usage of `aiokafka`:
```python
import aiokafka
import asyncio
import uuid


HEARTBEAT_INTERVAL = 0.1


def race_condition():
	loop = asyncio.get_event_loop()
	for i in range(1000):
		consumer = aiokafka.AIOKafkaConsumer(
			loop = loop,
			group_id = 'race_condition',
			enable_auto_commit = False,
			auto_offset_reset = 'earliest',
			heartbeat_interval_ms = HEARTBEAT_INTERVAL * 1000,
			metadata_max_age_ms = 30 * 1000,
		)
		topic = uuid.uuid4().hex
		print(i)
		loop.run_until_complete(run_test(consumer, topic))


async def run_test(consumer: aiokafka.AIOKafkaConsumer, topic):
	t1 = topic + '1'
	t2 = topic + '2'
	t3 = topic + '3'

	await consumer.start()
	consumer.subscribe([t1])
	await asyncio.sleep(HEARTBEAT_INTERVAL)

	consumer.subscribe([t1, t2])
	await asyncio.sleep(HEARTBEAT_INTERVAL / 10)
	consumer.subscribe([t1, t2, t3])
	await asyncio.sleep(HEARTBEAT_INTERVAL)

	consumer.subscribe([t2, t3])
	await asyncio.sleep(HEARTBEAT_INTERVAL)

	consumer.unsubscribe()

	await consumer.stop()


if __name__ == '__main__':
	race_condition()
```

When I run it with the fresh new kafka I can see the bug usually in few trying or at most 200 iterations. I tried using `await consumer._coordinator.ensure_active_group()` as @fabregas suggested but it did not help. Can you please take a look and try to replicate it too? I ran this with `aiokafka==0.2.0`.

Bellow is a sample of single run of this script (beware that I had to fix a bug while raising a `ValueError` in kafka-python `SubscriptionState.assign_from_subscribed()` https://github.com/dpkp/kafka-python/pull/960):
```
$ python race_condition.py 
0
1
2
3
Traceback (most recent call last):
  File "stuck/race_condition.py", line 48, in <module>
    race_condition()
  File "stuck/race_condition.py", line 21, in race_condition
    loop.run_until_complete(run_test(consumer, topic))
  File "/usr/lib/python3.5/asyncio/base_events.py", line 387, in run_until_complete
    return future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "stuck/race_condition.py", line 44, in run_test
    await consumer.stop()
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/consumer.py", line 292, in stop
    yield from self._coordinator.close()
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py", line 134, in close
    yield from self.heartbeat_task
  File "/usr/lib/python3.5/asyncio/futures.py", line 363, in __iter__
    return self.result()  # May raise too.
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py", line 795, in _heartbeat_task_routine
    yield from self.ensure_active_group()
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py", line 596, in ensure_active_group
    yield from self._perform_group_join()
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py", line 680, in _perform_group_join
    self.protocol, member_assignment_bytes)
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/group_coordinator.py", line 268, in _on_join_complete
    self._subscription.assign_from_subscribed(assignment.partitions())
  File "/home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/kafka/consumer/subscription_state.py", line 221, in assign_from_subscribed
    raise ValueError("Assigned partition %s for non-subscribed topic." % str(tp))
ValueError: Assigned partition TopicPartition(topic='53049f295c7d45c5b229684fed4655f51', partition=0) for non-subscribed topic.
Task was destroyed but it is pending!
task: <Task pending coro=<AIOKafkaClient._md_synchronizer() running at /home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/client.py:179> wait_for=<Future pending cb=[Task._wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending coro=<Fetcher._fetch_requests_routine() running at /home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/fetcher.py:250> wait_for=<Future pending cb=[Task._wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending coro=<AIOKafkaConnection._read() running at /home/sebekpet/.virtualenvs/messaging/lib/python3.5/site-packages/aiokafka/conn.py:154> wait_for=<Future pending cb=[Task._wakeup()]>>
```

**Comment by Drizzt1991** [comment]
Hey there, thanks for providing a way to reproduce it. I'm a bit overloaded atm, but will try to reproduce it on the week.
We did discuss this with @fabregas and it seems that it is quite hard to fix without changing the API of `subscribe` (basically it needs to be a coroutine). So the next minor version will probably include some API changes.

**Comment by Artimi** [comment]
Hello @Drizzt1991, 
I wanted to ask if you had a time to take a look? I would really like to see progress with this because one of our quite important services freezes several times per week due to this bug. What are the changes to do in aiokafka to prevent this bug?

**Comment by Drizzt1991** [comment]
@Artimi Hey there. Created a fix, that should help and merged into master. Is it possible to test it like that, or you need a release?

**Comment by Artimi** [comment]
@Drizzt1991 Thank you for the fix. I tried it with my script and also with the tests that causes the problem initially. It was both without an error for few hundreds repetition so I would consider the issue as fixed. When can I expect this to be released? Once again thanks for this, it will really help us.

**Comment by Drizzt1991** [comment]
Arround the end of the week i'd say. I want to merge everything to that time.

**Comment by Artimi** [comment]
Great, looking forward to it.

**Comment by Drizzt1991** [comment]
0.2.1 released

**Comment by Drizzt1991** [comment]
Hi there. While working on the release discovered another rare case of rebalancing race with `subscribe` - #118. If you do something like:
```python
consumer.subscribe([topic1])
await asyncio.sleep(0.5)
consumer.subscribe([topic2])
await consumer._coordinator.ensure_active_group()
```
it may break as consumer does not know yet about topic2's metadata and can assign 0 partitions by mistake . It can also race if you don't call `ensure_active_group()`, but not as bad. This may be another issue that can result in a stopped consumer, like you mentioned.
But as I sad it's a quite rare case, but if you happen to encounter it - ping me, I will think of a workaround, and I do plan on fixing this till 0.2.2.

**Comment by Drizzt1991** [comment]
The last race condition should be also fix as a result of #122. If any other problems are encountered lets open another issue. Thanks @Artimi for providing feedback on the issue 👍
